expression/builtin_op.go

```diff
@@ -391,6 +407,203 @@ func (b *builtinUnaryOpSig) eval(row []types.Datum) (d types.Datum, err error) {
 	return
 }
 
+type unaryMinusFunctionClass struct {
+	baseFunctionClass
+}
+
+func (b *unaryMinusFunctionClass) typeInfer(argExpr Expression, bf *baseBuiltinFunc) bool {
+	bf.tp.Init(mysql.TypeLonglong)
+	switch argExpr.GetType().Tp {
+	case mysql.TypeString, mysql.TypeVarchar, mysql.TypeVarString, mysql.TypeDouble, mysql.TypeFloat,
+		mysql.TypeDatetime, mysql.TypeDuration, mysql.TypeTimestamp:
+		bf.tp.Tp = mysql.TypeDouble
+	case mysql.TypeNewDecimal:
+		bf.tp.Tp = mysql.TypeNewDecimal
+	}
+	types.SetBinChsClnFlag(bf.tp)
+
+	overflow := false
+	if arg, ok := argExpr.(*Constant); ok {
```

Comment:
It's weird here. Can we change parser to parse `-9223372036854775809` as a constant decimal directly ?